third-party/realdds/src/dds-device-broadcaster.cpp

```diff
@@ -6,233 +6,154 @@
-                    {
-                        if( sn_and_handle.second.client_listener->_new_reader_joined )
-                        {
-                            if( send_device_info_msg( sn_and_handle.second ) )
-                            {
-                                sn_and_handle.second.client_listener->_new_reader_joined = false;
-                            }
-                        }
-                    }
-                } );
-            }
+            _stopped.set_trigger_value( true );
+            _th.join();
         }
-    } )
-{
-    if( ! _participant )
-        DDS_THROW( runtime_error, "dds_device_broadcaster called with a null participant" );
-}
-
-bool dds_device_broadcaster::run()
-{
-    if( _active )
-        DDS_THROW( runtime_error, "dds_device_broadcaster::run() called when already-active" );
-
-    _dds_device_dispatcher.start();
-    _new_client_handler.start();
-    _active = true;
-    return true;
-}
-
-void dds_device_broadcaster::add_device( device_info const & dev_info )
-{
-    if( ! _active )
-        DDS_THROW( runtime_error, "dds_device_broadcaster::add_device called before run()" );
-    // Post the connected device
-    handle_device_changes( {}, { dev_info } );
-}
-
-void dds_device_broadcaster::remove_device( device_info const & dev_info )
-{
-    if( ! _active )
-        DDS_THROW( runtime_error, "dds_device_broadcaster::remove_device called before run()" );
-    // Post the disconnected device
-    handle_device_changes( { dev_info.serial }, {} );
-}
-
-void dds_device_broadcaster::handle_device_changes( const std::vector< std::string > & devices_to_remove,
-                                                    const std::vector< device_info > & devices_to_add )
-{
-    // We want to be as quick as possible and not wait for DDS in any way -- so we handle
-    // device changes in the background, using invoke:
-    _dds_device_dispatcher.invoke( [this, devices_to_add, devices_to_remove]( dispatcher::cancellable_timer ) {
-        try
-        {
-            for( auto const & dev_to_remove : devices_to_remove )
-            {
-                remove_dds_device( dev_to_remove );
-            }
+    }
 
-            for( auto const & dev_to_add : devices_to_add )
-            {
-                if( ! add_dds_device( dev_to_add ) )
-                {
-                    LOG_ERROR( "Error creating a DDS writer" );
-                }
-            }
-        }
-        catch( ... )
+    std::shared_ptr< dds_topic_writer > register_broadcaster( dds_device_broadcaster * broadcaster )
+    {
+        // Each broadcaster (for a single device) gets its own writer, with a GUID, from which its broadcasts will be
+        // made. This lets the watcher associate the GUID with that specific device, and can tell when the GUID
+        // disappears that the device is no longer online...
+        auto writer = std::make_shared< dds_topic_writer >( _writer->topic(), _writer->publisher() );
```

Comment:
If the broadcaster is already in the set we don't need another writer

Reply:
Sure, but we don't call this function unless we're in a constructor for a broadcaster, so it's always new.